### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by bpteague** [cytoflow/cytoflow#286]
The original plan to keep `cytoflow` memory-efficient was this: store event (channel) data and per-event (condition) metadata in one big `pandas.DataFrame`. When you apply an `Operation` to an `Experiment`, make a shallow copy of `Experiment.data` and then either append a `Series` (if adding conditions eg with a `Threshold` op) or replace a Series (if modifying a channel, like with the `BleedthroughLinearOp`).

This plan was based on a fundamental misunderstanding of how `pandas` works. The `pandas` API presents a `DataFrame` as basically a list of `Series`. I figured that's how it worked under-the-hood -- you could just replace one Series with another without doing a deep copy of the entire underlying data set. That's not how it works - `pandas` is a (complex) wrapper for `numpy`, and it represents a 2D DataFrame as a `numpy.ndarray`. Can't just "refer" to part of it and change another part. AFAIK.

So ... what to do? Not yet clear. In the meantime, be explicit about it and make `Experiment.clone()` return a deep copy of `Experiment.data`.

**Comment by kristjaneerik** [comment]
Perhaps we could make a class that uses the setup you describe (the datastructure being a dictionary of `Series` object) that then implements (a minimal set) of key features of `DataFrame`s, like indexing, querying and `groupby` that are required for seaborn and anything you've used in cytoflow itself. The dictionary can then just hold references to unchanged things or change them to point to new `Series`.

I'd be happy to have a go at this -- or at least to at first see how much work this might be by investigating what functionality is absolutely necessary.

I haven't used much HDF5 + pandas, so I don't know without digging deeper, but there might be ways to use the pandas interface to HDF5 in ways that maintain shallow copies. There may even be built-in compression options in HDF5.

**Comment by kristjaneerik** [comment]
Another option would be to always append columns, never to replace/overwrite, and then each experiment would just have a view of the frame to the correct column subset. The column naming might get confusing / very long though.

**Comment by bpteague** [comment]
To be clear, I'm not talking about supporting datasets that don't fit in memory (: which is what I generally think of as being HDF5's use case. But if you've got a dataset that's a few hundred Mb big (not unreasonable), I would like to not make a full copy of it each time you call an operation's `apply()` function, because on a modest machine you quite rapidly run out of RAM.

And I suppose that we could re-jigger Experiment to be exactly what I initially thought a `DataFrame` was -- a collection of `pandas.Series`. The thing is, I *like* having the `DataFrame` be the underlying data structure. `Experiment.data` is a public attribute precisely because I want to enable people to use the rest of the scientific Python ecosystem on the underlying `DataFrame` when they need to do analyses that `cytoflow` doesn't support. In fact, my initial plan was to subclass `pandas.DataFrame` and add some flow-specific stuff -- but, for various reasons (mostly having to do with syntactic sugar that `pandas` provides) that was harder than it was worth.

Intuitively, I like the idea of each `Experiment` being a view of a specific set of columns from a larger append-only object. I'm not sure exactly how those semantics would work, though. Again, if only subclassing a `DataFrame` was easier....

**Comment by bpteague** [comment]
Thanks for reasoning through this with me, @kristjaneerik. I've learned more about `pandas` internals than I particularly wanted to, and it turns out that saying "`pandas` is a wrapper for `numpy.ndarray`" is pretty naive. Briefly:
- `pandas` data structures are built on 'blocks' of contiguous memory (`numpy.ndframe`s, mostly) managed by a class called a "block manager." 
- `DataFrame.copy(deep = False)` copies some of the metadata but not the index or the underlying blocks. Both the original `DataFrame` and the copy point back to the same blocks.
- Thus, operations that change the data in the blocks in-place on the *copy* will also update the *original.*
- Adding a new column adds new blocks to the block manager. So if a new column is added to the copy, it does NOT appear in the original.
- However, there are some operations that require making new copies of the underlying blocks -- if this happens to the copy, then subsequent changes to those blocks *won't* change the original.

For example:
```python
import pandas as pd

df2 = df.copy(deep = False)
s = pd.Series(range(7, 10), name = 'c')
df2.insert(2, 'c', s)

print(df, "\n")
print(df2, "\n")

# df2 has a new column, 'c', as expected, but df remains unchanged

df2['b'][0] = 10

print(df, "\n")
print(df2, "\n")

# when we change a value in df2['b'], the corresponding value in df changes, because both are using the same underlying block

df2['a'] = pd.Series(range(100, 103))

print(df, "\n")
print(df2, "\n")

# replacing a column in df2 does NOT replace the column in df

df2['b'][1] = 11

print(df, "\n")
print(df2, "\n")

# however, that seems to be because a copy has occurred, and changing a value in df2['b']
# is no longer reflected in df
```

It's not clear (and not documented AFAICT) which operations cause the block manager to copy blocks -- all of those details are hidden from the user.  However, these semantics actually match up pretty closely to what we want to do:
- Gating operations only add new columns to the `Experiment.data`. So doing a shallow copy here is fine; most of the data in the `DataFrame` continues to be shared with the previous `Experiment` instance.
- The operations that change the underlying data -- bleedthrough compensation, autofluorescence correction, etc -- are usually operating on ALL the fluorescence channels. So if most, or all, of the underlying blocks get copied as we update those columns, it's not a huge amount of extra memory.

I should be able to formalize these semantics in some extra tests -- I'll try to get to that in the next couple of days.

**Comment by bpteague** [comment]
Here's an excellent overview of the block manager and its machinations:
https://uwekorn.com/2020/05/24/the-one-pandas-internal.html

In the meantime, it looks like replacing a column only causes the block to get copied if it's of a different data type.  That's almost never the case in `cytoflow`. So, in ac4d91ef8fcdd6cc6c8b7101ccbb24afe27daa, I added a `deep` parameter to `Experiment.clone()`, defaults to `True`, that controls whether the copy of `Experiment.data` is deep or shallow. I also went through the operations and added `deep = False` where appropriate.  I think this is the best we can do for now; closing, unless @kristjaneerik has any other thoughts.
